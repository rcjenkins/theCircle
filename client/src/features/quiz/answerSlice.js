import { createSlice } from '@reduxjs/toolkit';

export const answersSlice = createSlice({
  name: 'answers',
  initialState: {
    answers: null,
    isFetching: false,
    error: null,
  },
  reducers: {
    requestSaveAnswers: (state) => {
      state.response = null;
      state.isFetching = true;
      state.error = null;
    },
    successSaveAnswers: (state, action) => {
      state.response = action.payload;
      state.isFetching = false;
      state.error = null;
    },
    failedSaveAnswers: (state, action) => {
      state.response = null;
      state.isFetching = false;
      state.error = action.payload.message;
    },
  },
});

// saving the answers

export const saveAnswers = () => (dispatch, getState) => {
  const {
    requestSaveAnswers,
    successSaveAnswers,
    failedSaveAnswers,
  } = answersSlice.actions;
  const { quiz: quizSlice } = getState();
  const { quiz, answers } = quizSlice;

  // wrapping the simple answers array for sending to API
  const payload = {
    quizData: {
      quizId: quiz.id.toString(),
      questionAnswers: answers,
    },
  };
  const apiEndpoint = window.awsAPI.apiEndpoint + '?quizId=' + quiz.id;
  dispatch(requestSaveAnswers());
  fetch(apiEndpoint, {
    method: 'POST',
    cache: 'no-cache',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(payload),
  })
    .then((response) => response.json())
    .then((json) => dispatch(successSaveAnswers(json)))
    .catch((error) => dispatch(failedSaveAnswers(error)));
};

export const selectAnswers = (state) => state.answers;

export default answersSlice.reducer;
