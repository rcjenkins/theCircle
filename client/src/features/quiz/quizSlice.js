import { createSlice } from '@reduxjs/toolkit';

export const quizSlice = createSlice({
  name: 'quiz',
  initialState: {
    answers: [],
    quiz: null,
    isFetching: false,
    error: null,
    currentQuestionId: null,
  },
  reducers: {
    requestLoadQuiz: (state) => {
      // Redux Toolkit allows us to write "mutating" logic in reducers. It
      // doesn't actually mutate the state because it uses the Immer library,
      // which detects changes to a "draft state" and produces a brand new
      // immutable state based off those changes
      state.answers = [];
      state.quiz = null;
      state.isFetching = true;
      state.error = null;
      state.currentQuestionId = null;
    },
    successLoadQuiz: (state, action) => {
      state.answers = [];
      state.quiz = action.payload;
      state.isFetching = false;
      state.error = null;
      // make the currentQuestionId the first question id after load as some quizzes may not start at 0
      state.currentQuestionId = state.quiz.questions[0].id;
    },
    failedLoadQuiz: (state, action) => {
      state.answers = [];
      state.quiz = null;
      state.isFetching = false;
      state.error = action.payload.message;
      state.currentQuestionId = null;
    },
    addAnswer: (state, action) => {
      state.answers.push({
        id: state.currentQuestionId.toString(),
        answers: action.payload.answers,
      });
      state.currentQuestionId = action.payload.nextQuestionId;
    },
    removeAnswer(state) {
      const lastAnswer = state.answers.pop();
      if (lastAnswer) {
        state.currentQuestionId = lastAnswer.id;
      }
    },
  },
});

// I like to hide away the actual actions this stops potential typos in the parameter object

export const addAnswer = (answers, nextQuestionId) => (dispatch) => {
  const { addAnswer } = quizSlice.actions;
  dispatch(addAnswer({ answers, nextQuestionId }));
};

export const removeAnswer = () => (dispatch) => {
  const { removeAnswer } = quizSlice.actions;
  dispatch(removeAnswer());
};

// loading quiz data using API

export const loadQuiz = () => (dispatch) => {
  const {
    requestLoadQuiz,
    successLoadQuiz,
    failedLoadQuiz,
  } = quizSlice.actions;
  dispatch(requestLoadQuiz());

  //const localEndpoint = `${window.location.origin}/quiz-0.json`;
  const apiEndpoint = window.awsAPI.apiEndpoint + '?quizId=0';
  fetch(apiEndpoint)
    .then((response) => response.json())
    .then((json) => dispatch(successLoadQuiz(json)))
    .catch((error) => dispatch(failedLoadQuiz(error)));
};

export const selectQuiz = (state) => state.quiz;

export default quizSlice.reducer;
