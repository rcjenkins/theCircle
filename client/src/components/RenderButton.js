import { Button, FormGroup } from 'reactstrap';
import React from 'react';

import { useSelector, useDispatch } from 'react-redux';
import { selectQuiz, addAnswer } from '../features/quiz/quizSlice';

const RenderButton = () => {
  const dispatch = useDispatch();
  const quizData = useSelector(selectQuiz);
  const question = quizData.quiz.questions.find(
    (element) => element.id === quizData.currentQuestionId,
  );

  const handleClick = (id) => {
    const nextPage = question.answers.find((element) => element.id === id)
      .nextQuestionId;
    dispatch(addAnswer([id.toString()], nextPage));
  };

  const button = (answer) => {
    return (
      <React.Fragment key={answer.id}>
        <Button
          aria-label="question"
          id={`button-${answer.id}`}
          color="primary"
          onClick={() => {
            handleClick(answer.id);
          }}
        >
          {answer.text}
        </Button>{' '}
      </React.Fragment>
    );
  };

  return (
    <FormGroup role="group">
      <div id="question" className="label label-default">
        {question.title}
      </div>
      <span>
        {question.answers.map((answer) => {
          return button(answer);
        })}
      </span>
    </FormGroup>
  );
};

export default RenderButton;
