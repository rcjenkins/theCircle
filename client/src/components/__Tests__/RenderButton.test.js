import React from 'react';
import RenderButton from '../RenderButton';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { selectQuiz } from '../../features/quiz/quizSlice';

jest.mock('../../features/quiz/quizSlice');

it('RenderButton renders correctly', () => {
  selectQuiz.mockReturnValue({
    answers: [],
    isFetching: false,
    error: null,
    currentQuestionId: 0,
    quiz: {
      id: 0,
      title: 'Daily Engagement Service',
      questions: [
        {
          id: 0,
          title: 'Did you enjoy yourself today?',
          type: 'button',
          answers: [
            {
              id: 0,
              text: 'Hell Yes!',
              nextQuestionId: 1,
            },
            {
              id: 1,
              text: 'Nope... :-(',
              nextQuestionId: 2,
            },
          ],
        },
      ],
    },
  });

  const middlewares = [thunk];
  const mockStore = configureStore(middlewares);

  const initialState = {};
  const store = mockStore(initialState);
  const tree = renderer
    .create(
      <Provider store={store}>
        <RenderButton />
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
