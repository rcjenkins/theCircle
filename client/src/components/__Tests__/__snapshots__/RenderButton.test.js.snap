// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderButton renders correctly 1`] = `
<div
  className="form-group"
  role="group"
>
  <div
    className="label label-default"
    id="question"
  >
    Did you enjoy yourself today?
  </div>
  <span>
    <button
      aria-label="question"
      className="btn btn-primary"
      id="button-0"
      onClick={[Function]}
      type="button"
    >
      Hell Yes!
    </button>
     
    <button
      aria-label="question"
      className="btn btn-primary"
      id="button-1"
      onClick={[Function]}
      type="button"
    >
      Nope... :-(
    </button>
     
  </span>
</div>
`;
